### PR TITLE
*: check visibility for show databases and tables

### PR DESCRIPTION
**Pull request by tiancaiamao**
`show databases` and `show tables` return just those visible to the user,
if the user doesn't have privilege, he should not see it.

### CODE REVIEW

File: executor/show.go [hunk 1]
```diff
@@ -179,9 +183,14 @@ func (e *ShowExec) fetchShowTables() error {
 	if !e.is.SchemaExists(e.DBName) {
 		return errors.Errorf("Can not find DB: %s", e.DBName)
 	}
+	checker := privilege.GetPrivilegeChecker(e.ctx)
 	// sort for tables
 	var tableNames []string
 	for _, v := range e.is.SchemaTables(e.DBName) {
+		// TODO: Should consider column privileges, which also make a table visible.
+		if checker != nil && !checker.RequestVerification(e.DBName.O, v.Meta().Name.O, "", mysql.AllPriv-1) {
```
Comment (shenli) [inline executor/show.go]: What's mean by `mysql.AllPriv-1`?

Comment (tiancaiamao) [inline executor/show.go]: All privilege bits set to 1, means any kind of privileges would be OK.

**Comment by shenli** [comment]
LGTM
@coocood @zimulala PTAL

File: privilege/privileges/privileges.go [hunk 2]
```diff
@@ -246,6 +246,31 @@ func (p *UserPrivileges) ConnectionVerification(user, host string, auth, salt []
 	return true
 }
 
+// ShowDatabase implements the Checker interface.
+func (p *UserPrivileges) DBVisible(db string) bool {
```
Comment (coocood) [inline privilege/privileges/privileges.go]: s/DBVisible/DBIsVisible

**Comment by zimulala** [comment]
LGTM

File: privilege/privileges/cache.go [hunk 3]
```diff
@@ -418,6 +418,43 @@ func (p *MySQLPrivilege) RequestVerification(user, host, db, table, column strin
 	return false
 }
 
+// ShowDatabase checks whether the user can see the db.
+func (p *MySQLPrivilege) ShowDatabase(user, host, db string) bool {
```
Comment (coocood) [inline privilege/privileges/cache.go]: DBIsVisible

**Comment by coocood** [comment]
LGTM

**Pull request merged**